linux/src/async_socket_linux.c

```diff
@@ -0,0 +1,695 @@
+                                receive_result = ASYNC_SOCKET_RECEIVE_OK;
+                                break;
+                            }
+                            else
+                            {
+                                index++;
+                            }
+                        }
+                    } while (true);
+                    // Call the callback
+                    recv_context->on_receive_complete(recv_context->on_receive_complete_context, receive_result, recv_size);
+
+                    // Free the memory
+                    free(recv_context);
+                }
+            }
+            else if (events[index].events & EPOLLOUT)
+            {
+                ASYNC_SOCKET_SEND_CONTEXT* send_context = events[index].data.ptr;
+
+                uint32_t index;
+                for (index = 0; index < send_context->total_buffer_count; index++)
+                {
+                    ssize_t send_size = 0;
+                    ASYNC_SOCKET_SEND_RESULT send_result = ASYNC_SOCKET_SEND_OK;
+                    // loop here to send all the bytes
+                    do
+                    {
+                        void* send_pos = send_context->buffers[index].buffer + send_size;
+                        uint32_t send_length = send_context->buffers[index].length - send_size;
+                        send_size += send(send_context->socket_handle, send_pos, send_length, 0);
+                        if (send_size < 0 && (errno != EAGAIN && errno != EWOULDBLOCK) )
+                        {
+                            if (errno == ECONNRESET)
+                            {
+                                send_result = ASYNC_SOCKET_SEND_ABANDONED;
+                            }
+                            else
+                            {
+                                send_result = ASYNC_SOCKET_SEND_ERROR;
+                                // Log Error here
+                                LogError("failure sending data length: %" PRIu32 "", send_length);
+                                break;
+                            }
+                        }
+                    } while (send_size < send_context->buffers[index].length);
+
+                    send_context->on_send_complete(send_context->on_send_complete_context, send_result);
+                }
+                free(send_context);
+            }
+        }
+    } while (interlocked_add(&g_threads_num, 0) > 0);
+    return 0;
+}
+
+static int initialize_global_thread(void)
+{
+    int result = -1;
+
+    int32_t current_count = interlocked_increment(&g_threads_num);
+    if (current_count <= THREAD_COUNT)
+    {
+        g_epoll = epoll_create(MAX_EVENTS_NUM);
+        if (g_epoll == -1)
+        {
+            LogError("failure epoll_create error num: %d", errno);
+        }
+        else
+        {
+            result = g_epoll;
+            for (int32_t index = 0; index < THREAD_COUNT; index++)
```

Comment:
>index < THREAD_COUNT [](http://example.com/codeflow?start=36&length=20)

Ah, I see, you are creating multiple threads with 64 events each.

Reply:
Or not.
THREAD_COUNT is 1. Soooo this won;t work for more than 64 sockets.

Reply:
The MAX_EVENTS_NUM is ignored since kernel 2.6.8, but on on earlier versions this is the note: 

The kernel used this information as a hint for the amount of space to initially allocate in internal data structures describing events

So technically I could just put 1 and it would figure it out.

[https://man7.org/linux/man-pages/man2/epoll_create.2.html](https://man7.org/linux/man-pages/man2/epoll_create.2.html)